File: credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(void)
{
    long cardNumber;
    do
    {
        cardNumber = get_long("Number: ");
        // printf("%li\n", cardNumber);
    }
    while (cardNumber < 0);
    
    //counts if no of digits == 13,15,or 16
    int count = 0; 
    long digits = cardNumber; //counting n_digits in cardnumber
    while (digits > 0)
    {
        digits = digits / 10;
        count++;
    }
    if ((count != 13) && (count != 15) && (count != 16))
    {
        printf("INVALID\n");
        return 0;
    }
    
    //initialising number array with the cardnumbers reversed
    int number[count];
    for (int i = 0; i < count; i++)
    {
        number[i] = cardNumber % 10; //reverses the cardnumber
        cardNumber = cardNumber / 10;
    }

    //doubling the cardnumbers, two steps from the second to the last digit
    int even_e;
    int total = 0;
    
    for (int i= 1; i < count; i+=2)
    {
        even_e = 2 * number[i]; //doubles and adds
        if (even_e > 9)
        {
            int firstDigit = even_e /  10;
            int lastDigit = even_e % 10;
            total = total + firstDigit + lastDigit;
        }
        else
        {
            total  = total + even_e;
        }
    }

    for (int j = 0; j < count; j+=2)
    {
        total = total + number[j];
    }
    
    //checking  validity of card
    if (count == 15)
    {
        if (total % 10 == 0 && number[14] == 3 && (number[13] == 4 || number[13] == 7))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    else if (count == 16)
    {
        if (total % 10 == 0 && number[15] == 5 && (number[14] == 1 || number[14] == 2 || number[14] == 3 || number[14] == 4 || number[14] == 5))
        {
            printf("MASTERCARD\n");
        }
        else if ((total % 10 == 0 && number[15] == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
        
    }
    
    else if (count == 13)
    {
        if (total  % 10 == 0 && number[12] == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }       
    }
    return 0;
}

File: credit/credit.py
cardNumber = input("Card Number: ")

#checking user input
while True:
    if (cardNumber.isalpha()) or (cardNumber == '') or ("-" in cardNumber) or (int(cardNumber) < 0) :
        cardNumber = input("Card Number: ")
    else: break

#functions
def checkSumValidCard(cardNumber):
    """
    This checksums for digits in the card number and prints if the card is valid or not
    """
    total = 0
    cardNumber.reverse() #reverses the list
    
    #Visa= 13,16, American Express = 16, MasterCard = 15
    if len(cardNumber) not in [13,15,16]:
        print('INVALID')
#         print("Please input a valid  card")
    else:
        if len(cardNumber) in [13, 15, 16]:
            for _ , num_e in enumerate(cardNumber[1::2],1): #takes from the second to the last digit
                even = 2 * int(num_e)
                if even > 9:
                    lastDigit = even % 10
                    firstDigit = even // 10
                    total = total + lastDigit + firstDigit
                else:
                    total = total + even
            for _, num_o in enumerate(cardNumber[::2], 1):
                total = total + int(num_o)
        
        #checking valid card
        cardNumber.reverse() #reversing back to original
        if total % 10 == 0 and int(str(cardNumber[0]) + str(cardNumber[1])) in [34, 37]: print("AMERICAN EXPRESS")
        elif total % 10 == 0 and  int(str(cardNumber[0]) + str(cardNumber[1])) in [51,52,53,54,55]: print("MASTERCARD")
        elif total % 10 == 0 and int(str(cardNumber[0])) == 4: print("VISA")
        else: print("INVALID")


if __name__ == "__main__":
    cardNumber = [int(x) for x in cardNumber]
    checkSumValidCard(cardNumber)
